File: darts-suite/fortran/darts-hybrid.f
! Compute pi hybrid MPI/OpenMP
        program darts
          use lcgenerator
          implicit none
          include 'mpif.h'
          integer :: num_trials = 1000000, i = 0, Ncirc = 0
          real :: pi = 0.0, x = 0.0, y = 0.0, r = 1.0
          real :: r2 = 0.0

          integer :: rank, np, manager = 0
          integer :: provided, mpistatus, mpierr, j
          integer :: my_trials, temp

          call MPI_Init_thread(MPI_THREAD_MULTIPLE, provided, mpierr)
          call MPI_Comm_size(MPI_COMM_WORLD, np, mpierr)
          call MPI_Comm_rank(MPI_COMM_WORLD, rank, mpierr)
          r2 = r*r
          my_trials = num_trials/np
          if (mod(num_trials, np) .gt. rank) then
            my_trials = my_trials+1
          end if
          call seed(rank)

!$OMP parallel private(x,y) reduction(+:Ncirc)
!$OMP do 
          do i = 1, my_trials
!$OMP critical (randoms)
            x = lcgrandom()
            y = lcgrandom()
!$OMP end critical (randoms)
            if ((x*x + y*y) .le. r2) then
              Ncirc = Ncirc+1
            end if
          end do
!$OMP end parallel

          call MPI_Reduce(Ncirc, temp, 1, MPI_INTEGER, MPI_SUM,
     &      manager, MPI_COMM_WORLD, mpierr)

          if (rank .eq. manager) then
              Ncirc = temp
              pi = 4.0*((1.0*Ncirc)/(1.0*num_trials))
              print*, '     '
              print*, '     Computing pi using hybrid MPI/OpenMP:'
              print*, '     For ', num_trials, ' trials, pi = ', pi
              print*, '     '
          end if
          call MPI_Finalize(mpierr)
        end

File: darts-suite/c/darts-hybrid.c
/* Compute pi using hybrid MPI/OpenMP */
#include "lcgenerator.h"
#include <mpi.h>
#include <omp.h>
#include <stdio.h>

int main(int argc, char **argv) {

    long num_trials = 1000000;  /* Statically initialized on all processes */
    long total_Ncirc = 0;
    long local_rank_Ncirc = 0;
    double pi;
    double r = 1.0;
    double r2 = r*r;
    long base_seed = 12345;

    int rank, size, manager = 0;
    int total_threads = 1;
    int provided;

    MPI_Init_thread(&argc, &argv, MPI_THREAD_FUNNELED, &provided);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    #pragma omp parallel
    {
        #pragma omp single
        total_threads = omp_get_num_threads();
    }

    long trials_per_rank = num_trials / size;
    long trials_per_thread = trials_per_rank / total_threads;

    #pragma omp parallel reduction(+:local_rank_Ncirc)
    {
        int tid = omp_get_thread_num();
        double x, y;

        for (long i = 0; i < trials_per_thread; i++) {
            long global_i = (rank * total_threads) + tid + (i * (size * total_threads));
            long thread_seed = base_seed + (global_i * 2L);

            x = lcgrandom_r(&thread_seed);
            y = lcgrandom_r(&thread_seed);

            if ((x*x + y*y) <= r2) {
                local_rank_Ncirc++;
            }
        }
    }

    MPI_Reduce(&local_rank_Ncirc, &total_Ncirc, 1, MPI_LONG, MPI_SUM, manager, MPI_COMM_WORLD);

    if (rank == manager) {
        pi = 4.0 * ((double)total_Ncirc) / ((double)num_trials);
        printf("\n \t Computing pi using hybrid MPI/OpenMP: \n");
        printf("\t For %ld trials, pi = %f\n", num_trials, pi);
        printf("\n");
    }

    MPI_Finalize();
    return 0;
}
